File: python/713.py
"""Project Euler Problem 713: Turán Graphs.

If there are N fuses and m are working, let T(N, m) be the minimum number of
pairs of fuses to ensure that at least one pair of fuses are both working.
Find Σ_{m=2}^N T(N, m).

T(N, m) is equivalent to constructing a graph over N vertices with the minimum
number of edges such that any subset of m vertices will contain an edge. This
can only be done by splitting the vertices into k = m-1 groups, with an edge
connecting all vertices in a group, because by the Pigeonhole principle at
least two of the m vertices will be in the same group.

To minimize the number of edges, we need to make the sizes of the groups as
similar as possible, i.e. N % k have N/k + 1 vertices, and the other groups
have N/k vertices. The number of edges is then (N % k) nCr(N/k + 1, 2) +
(k - N % k) nCr(N/k, 2). This is the Turan graph and is the minimum by
Turan's theorem.
"""

from __future__ import annotations


def ncr(n: int, r: int) -> int:
    """Binomial coefficient C(n, r)."""
    if r < 0 or r > n:
        return 0
    if r == 0 or r == n:
        return 1
    result = 1
    for i in range(min(r, n - r)):
        result = result * (n - i) // (i + 1)
    return result


def solve() -> int:
    """Solve Problem 713."""
    n = 10**7
    ans = 0

    for k in range(1, n):
        group_size = n // k
        remainder = n % k
        ans += remainder * ncr(group_size + 1, 2) + (k - remainder) * ncr(
            group_size, 2
        )

    return ans


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/713.c
/*
 * Project Euler Problem 713: Turan Graphs.
 *
 * T(N, m) = (N%k) * C(floor(N/k)+1, 2) + (k - N%k) * C(floor(N/k), 2)
 * where k = m - 1. Sum over m = 2..N, i.e., k = 1..N-1.
 *
 * Optimized: group by floor(N/k) values using floor quotient grouping.
 */
#include <stdio.h>

typedef long long ll;

int main() {
    ll n = 10000000LL;
    ll ans = 0;

    for (ll k = 1; k < n; k++) {
        ll gs = n / k;       /* group_size = floor(n/k) */
        ll rem = n % k;      /* remainder = n mod k */
        /* rem * C(gs+1, 2) + (k - rem) * C(gs, 2)
         * = rem * gs*(gs+1)/2 + (k - rem) * gs*(gs-1)/2
         * = gs/2 * (rem*(gs+1) + (k-rem)*(gs-1))
         * = gs/2 * (rem*gs + rem + k*gs - k*rem - rem*gs + rem + k*rem - rem)
         * Hmm, let's just compute directly.
         */
        ll c1 = gs * (gs + 1) / 2;  /* C(gs+1, 2) */
        ll c2 = gs * (gs - 1) / 2;  /* C(gs, 2) */
        ans += rem * c1 + (k - rem) * c2;
    }

    printf("%lld\n", ans);
    return 0;
}
